### PR TITLE
storage: refactor kafka sinks (new attempt)

### PR DESCRIPTION
**Pull request by petrosagg**
### Motivation

This PR is re-adding the original kafka sink refactor. The only modifications are contained in the last commit which adjust the parsing of the legacy progress messages to handle the empty frontier case which was added in https://github.com/MaterializeInc/materialize/pull/24182/commits/7204ebed9bdfc9eb0ef890e79572c69ac03f17c1.

### Tips for reviewer

Only look at the last commit

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by umanwizard** [comment]
Removing myself since according to @petrosagg , the Adapter bits have already been reviewed (by @jkosh44 in #23790)

File: src/storage-client/src/sink.rs [hunk 1]
```diff
@@ -327,7 +327,19 @@ pub async fn publish_kafka_schemas(
...
+    pub timestamp: Option<Option<Timestamp>>,
+}
+
+// Any value that is present is considered Some value, including null.
+fn deserialize_some<'de, T, D>(deserializer: D) -> Result<Option<T>, D::Error>
+where
+    T: Deserialize<'de>,
+    D: Deserializer<'de>,
+{
+    Deserialize::deserialize(deserializer).map(Some)
```
Comment (guswynn) [inline src/storage-client/src/sink.rs:342]: this is a bit terrifying but thanks for the link!

**Review by guswynn — APPROVED** [review]
helll ya, lets try it

File: src/storage-client/src/sink.rs (hunk 1 above)
Comment (petrosagg) [reply src/storage-client/src/sink.rs:342]: `dtolnay` wrote it, it's from The Book!

**Pull request merged**